Fix issue PLAT-452: Draft block showing as published.

doctoryes opened the pull request:
The previous logic was making draft units show up as published.

doctoryes commented:
@adampalay  This change fixes the branch unit showing up as published. However, other tests are failing - I'm looking into those tests now.

doctoryes commented:
https://openedx.atlassian.net/browse/PLAT-452

doctoryes commented:
@cpennington @adampalay  I have not yet written a test for the original issue - it's obviously missing from our unit tests. And I don't consider this PR done until that test is written. (Or maybe a separate PR, depending on how quickly we want to release this.)

However, this PR is ready for an initial review.

cpennington commented on common/lib/xmodule/xmodule/modulestore/__init__.py:
```diff
@@ -279,6 +279,69 @@ def _is_in_bulk_operation(self, course_key, ignore_case=False):
+        Setter for the value. Also marks the value as valid.
+        """
+        self._value = value
+        self.valid = True
```
A pattern we've used for this in other places is somewhat simpler:

    UNSET = object()  # This guarantees that it UNSET == UNSET, but doesn't == anything else

And then you can check `is UNSET` to find out if a value is unset or not.

doctoryes commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -15,7 +15,9 @@
+# This import breaks this test file when run separately. The test is currently skipped below -
+# so remove this import for now. (PLAT-449)
+#from mock_django import mock_signal_receiver
 from nose.plugins.attrib import attr
```
@clintonb  FYI.

adampalay commented on common/lib/xmodule/xmodule/modulestore/__init__.py:
```diff
@@ -279,6 +279,69 @@ def _is_in_bulk_operation(self, course_key, ignore_case=False):
+        self.value = value
+
+    def __str__(self):
+        return "SourceVersion(version={0.version}, valid={0.valid}".format(self)
```
missing a close parenthesis

adampalay commented on common/lib/xmodule/xmodule/modulestore/split_mongo/caching_descriptor_system.py:
```diff
@@ -242,7 +242,7 @@ def xblock_from_json(self, class_, course_key, block_key, block_data, course_ent
         module.previous_version = edit_info.previous_version
         module.update_version = edit_info.update_version
-        module.source_version = edit_info.source_version
+        module.source_version = edit_info._source_version.version  # pylint: disable=protected-access
```
Running this branch locally, I'm running into an error on this line, because `edit_info` doesn't have a `_source_version` attribute. Can we set a default value for this, if it's not set?

doctoryes commented:
test this please

doctoryes commented:
test this please

doctoryes commented on common/lib/xmodule/xmodule/modulestore/xml_exporter.py:
```diff
@@ -153,20 +153,15 @@ def export(self):
 
             # export only the published content
             with self.modulestore.branch_setting(ModuleStoreEnum.Branch.published_only, self.courselike_key):
+                courselike = self.get_courselike()
```
The above was the key bug here. The course was being retrieved outside the `branch_setting` context manager that restricted course items to published items only.

doctoryes commented:
test this please

cpennington commented on common/lib/xmodule/xmodule/modulestore/__init__.py:
```diff
@@ -342,16 +360,32 @@ def from_storable(self, edit_info):
                 "original_usage_version={0.original_usage_version}, "
                 "_subtree_edited_on={0._subtree_edited_on}, "
-                "_subtree_edited_by={0._subtree_edited_by})").format(self)
+                "_subtree_edited_by={0._subtree_edited_by})").format(
```
Optional: These might be clearer if they were written with keyword arguments in the format string.

cpennington replied:
Nit: `EditInfo` should be written as `0.__class__.__name__`, so that it covers subclasses correctly.

cpennington commented on common/lib/xmodule/xmodule/modulestore/__init__.py:
```diff
@@ -398,7 +432,7 @@ def from_storable(self, block_data):
 
-    def __str__(self):
+    def __repr__(self):
         return ("BlockData(fields={0.fields}, "
```
Nit: `BlockData` should be `0.__class__.__name__`.

cpennington commented on common/lib/xmodule/xmodule/modulestore/split_mongo/caching_descriptor_system.py:
```diff
@@ -242,7 +242,7 @@ def xblock_from_json(self, class_, course_key, block_key, block_data, course_ent
         module.previous_version = edit_info.previous_version
         module.update_version = edit_info.update_version
-        module.source_version = edit_info.source_version
+        module.source_version = edit_info.get_source_version(None)
```
Optional: this would be clearer if passed as a keyword argument.

cpennington commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -506,6 +512,213 @@ def test_has_changes(self, default_ms):
+                )
+
+                # Verify that the imported block still is a draft, i.e. has changes.
+                self.assertTrue(self._has_changes(draft_xblock.location))
```
Optional: It might be worth checking the contents of the block fields, as well, to verify that those haven't changed.

cpennington commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -506,6 +512,213 @@ def test_has_changes(self, default_ms):
+                # Verify that the published block still has a draft block, i.e. has changes.
+                self.assertTrue(self._has_changes(chapter.location))
+                self.assertTrue(self._has_changes(sequential.location))
+                self.assertTrue(self._has_changes(vertical.location))
```
Optional: Would it be worth extracting the "set up data, export, import, compare" pattern as a method or a test class?

doctoryes commented:
@cpennington I've addressed all your comments except comparing block fields after import. I've also squashed.

@adampalay @Shrhawk @waheedahmed  This issue (PLAT-452) has apparently become a CAT-1 so I'm looking for some reviews today.

cpennington commented on common/lib/xmodule/xmodule/modulestore/__init__.py:
```diff
@@ -342,16 +360,40 @@ def from_storable(self, edit_info):
+            original_usage_version=self.original_usage_version,
+            subtree_edited_on=self._subtree_edited_on,
+            subtree_edited_by=self._subtree_edited_by,
+        )  # pylint: disable=bad-continuation
```
Nit: I think it was actually more readable when you passed `self` in, and accessed the attributes in the format strings. I just thought that rather than using `{0.some_attribute}`, `{self.some_attribute}` would be clearer.

cpennington commented on common/lib/xmodule/xmodule/modulestore/__init__.py:
```diff
@@ -398,13 +440,21 @@ def from_storable(self, block_data):
+            definition_loaded=self.definition_loaded,
+            defaults=self.defaults,
+            edit_info=self.edit_info,
+        )
```
Nit: As above.

cpennington commented:
:+1:

adampalay commented:
I'll review this. Hopefully we can put it out as a hotfix once the release goes out

doctoryes commented:
test this please

adampalay commented:
@doctoryes , we don't have expected behavior in the following scenario:
1) Go to split mongo course
2) Create html component
3) Publish component
4) Make changes to the component so it becomes a draft again
5) export course
6) import course

Expected behavior: we see the draft version of the html unit as draft
Actual behavior: we see the draft version of the html unit as published

Shrhawk commented:
@doctoryes i didn't get your changes here, can you give the summary of the this pr ?

doctoryes commented:
@adampalay Did this newly reported problem start occurring on this branch? Does it work properly on master?

adampalay commented:
@doctoryes I think it was always an issue, but let me verify

adampalay commented:
@doctoryes , yes, this problem occurs on master too

doctoryes commented on common/lib/xmodule/xmodule/modulestore/tests/test_mixed_modulestore.py:
```diff
@@ -506,6 +513,232 @@ def test_has_changes(self, default_ms):
+
+    @ddt.data(ModuleStoreEnum.Type.mongo, ModuleStoreEnum.Type.split)
+    def test_vertical_with_draft_and_published_unit_has_changes_before_export_and_after_import(self, default_ms):
+        """
```
@cpennington  This is the test that repros the current issue.

doctoryes commented on common/lib/xmodule/xmodule/modulestore/xml_importer.py:
```diff
@@ -419,7 +419,8 @@ def run_imports(self):
-            with self.store.bulk_operations(dest_id):
+            #with self.store.bulk_operations(dest_id):
+            if True:
                 source_courselike, courselike, data_path = self.get_courselike(courselike_key, runtime, dest_id)
```
@cpennington  And this is the hack "fix" for the problem - I'd like to discuss this with you.

cpennington replied:
Yeah, let's talk. I'd love to understand why using `bulk_operations` breaks things (and figure out how to fix it).

doctoryes commented:
@adampalay @cpennington @Shrhawk - This is ready for a final review. I've added comments to each individual commit describing it (if needed).

@andy-armstrong - This issue was Studio-related. A review of my added tests to see if my assumptions are correct would be welcome.

andy-armstrong commented on common/lib/xmodule/xmodule/modulestore/split_mongo/split_draft.py:
```diff
@@ -500,7 +493,7 @@ def import_xblock(self, user_id, course_key, block_type, block_id, fields=None,
                 with self.branch_setting(ModuleStoreEnum.Branch.draft_preferred, draft_course):
                     draft_block = self.import_xblock(user_id, draft_course, block_type, block_id, fields, runtime)
-                    return self.publish(draft_block.location.version_agnostic(), user_id, blacklist=EXCLUDE_ALL, **kwargs)
+                    # Don't publish the draft block here. Instead create the published block below.
```
I find this logic hard to understand now (although, to be fair the old logic was confusing too!). So this path is for 'published_only', but it no longer directly publishes the draft but instead it now continues on to do the regular import. It seems that it is now doing more work for 'published_only' than it does for other branch settings, which doesn't seem right intuitively. Could you add some clarifying comments explaining how this logic works in the regular path, and then what is different for ``published_only``.

adampalay replied:
this is also confusing to me. Why were we publishing drafts during imports in the first place?

andy-armstrong commented:
:+1: the tests all look good, so my one caveat is that I don't understand the import change in split_draft. An explanation in a comment would be sufficient, assuming that the logic is correct.

Shrhawk commented:
@doctoryes this seems complicated to me !  Can you update the first comment with logic ?

adampalay commented:
@doctoryes , unfortunately, I discovered another "buggy" scenario:
1) create html component
2) publish component
3) export course
4) import course
expected: component is published
actual: component is in draft (although seemingly nothing has changed about it)

cpennington commented:
:+1:, but I would like to see more assertions on field-values during the import/export cycles, to make sure that the blocks that have changes have the changes we expect (and that the ones that don't don't)

doctoryes commented:
@cpennington This is ready for another review - I've made enough changes that you should likely look at all of it.

@adampalay Feel free to uncover the next issue. :8ball: And a review would be grand.

@Shrhawk Have I added enough comments for my changes to be understandable? I'll add more if needed.

adampalay commented on common/lib/xmodule/xmodule/modulestore/xml_exporter.py:
```diff
@@ -42,6 +42,9 @@ def _export_drafts(modulestore, course_key, export_fs, xml_centric_course_key):
         )
+        # Check to see if the returned draft modules have changes w.r.t. the published module.
+        # Only modules with changes will be exported into the /drafts directory.
+        draft_modules = [module for module in draft_modules if modulestore.has_changes(module)]
```
awesome

cpennington replied:
It seems like, given the `revision=ModuleStoreEnum.RevisionOption.draft_only` qualifier on line 43, any modulestore that returns non-draft content (content where has_changes is false) is misbehaving. Although, I guess Studio allows for drafts that don't differ in content from the published version (that's what you get when you first hit "Edit draft" in Studio). Blarg. Such unclear semantics.

adampalay commented:
:+1: !

cpennington commented:
:+1:

doctoryes commented:
Rebased this against the current RC as this PR:

https://github.com/edx/edx-platform/pull/7910

adampalay commented:
@feanil , why did you merge this? Aren't these commits already in the RC?

feanil commented:
@adampalay it got merged as part of the release merging back to master.
